x-exception/src/X/Control/Monad/Catch.hs

```diff
@@ -43,23 +44,25 @@ bracketF a f g =
 --
 -- Exception and `Left` safe version of bracketEitherT.
 --
-bracketEitherT' :: MonadMask m => EitherT e m a -> (a -> EitherT e m c) -> (a -> EitherT e m b) -> EitherT e m b
+bracketEitherT'
+  :: forall m e a b c. MonadMask m
+  => EitherT e m a
+  -> (a -> EitherT e m c)
+  -> (a -> EitherT e m b)
+  -> EitherT e m b
 bracketEitherT' acquire release run =
-  EitherT $ bracketF
-    (runEitherT acquire)
-    (\r -> case r of
-      Left _ ->
-        -- Acquire failed, we have nothing to release
-        return . Right $ ()
-      Right r' ->
-        -- Acquire succeeded, we need to try and release
-        runEitherT (release r') >>= \x -> return $ case x of
-          Left err -> Left (Left err)
-          Right _ -> Right ())
-    (\r -> case r of
-      Left err ->
-        -- Acquire failed, we have nothing to run
-        return . Left $ err
-      Right r' ->
-        -- Acquire succeeded, we can do some work
-        runEitherT (run r'))
+  let
+    acquire' :: m (Either e a)
+    acquire' = runEitherT acquire
+
+    release' :: Either e a -> m (Either e c)
+    release' = either (return . Left) (runEitherT . release) -- if acquire' failed then pass through
+
+    run' :: Either e a -> m (Either e b) 
+    run' = either (return . Left) (runEitherT . run) -- if acquire' failed then skip run.
+
+ in EitherT $ mask $ \unmask -> do
+  resource <- acquire'
+  result <- unmask (run' resource) `onException` release' resource
+  releaseResult <- release' resource
```

Comment:
I could easily be missing something, but this looks like release isn't going to be run if run has a value level fail (i.e. Left). I don't think there are any redundant cases that can be eliminated in the original version, if you want to try to clean it up, you could replace the nested either usage with a flattened out data-type, might help?